README.md
PyPredict
=======

Satellite tracking and pass prediction library for Python.

PyPredict is a C Python extension directly adapted from the ubiquitous [predict](http://www.qsl.net/kd2bd/predict.html) satellite tracking tool.
We aim for result-parity and it should produce identical values on the same inputs.

If you think you've found an error, please include predict's differing output in the bug report.
If you think you've found a bug in predict, please report and we'll coordinate with upstream.

## Installation
```
sudo python setup.py install
```
## Usage
```
predict.predict(epoch, tle, qth)
# \_ epoch - number of seconds since epoch (as float)
# \_ tle - array of strings (lines 0,1,2 of TLE)
# \_ qth - (optional) tuple of latitude (float), longitude (float), altitude (int)
```
## Example
```
import time
import urllib2
import predict

# Get a TLE
res = urllib2.urlopen("http://tle.nanosatisfi.com/%d" % 40044)
if res.getcode() != 200:
    raise Exception("Unable to retrieve TLE from tle.nanosatisfi.com. HTTP code(%s)", res.getcode())
tle = res.read().rstrip()

predict.predict(time.time(), tle.split('\n'), (37.7727, 122.407, 25))
```

CHANGES
Release 2.2.3:
By John A. Magliacane <kd2bd@amsat.org> (15-Apr-2006):

* PREDICT now handles -u, -t, and -q command-line options
  (hopefully) better than before.  When these options are used
  and no default TLE and QTH files are found under the user's
  home directory, PREDICT is no longer placed into "new user"
  mode.  This should ease the use of PREDICT in non-interactive
  applications (in automated scripts, etc.) where a $HOME
  environment is lacking or not desired.  Only if a user
  has no QTH and TLE files under their home directory, and
  has passed no -q and -t arguments to PREDICT will that
  user now be considered a "new user".

* New command-line options were added to PREDICT.  The -east
  option causes PREDICT to display longitudes in degrees east
  rather than in degrees west.  The -south option displays
  latitudes in degrees south rather than in degrees north.
  -west and -north options, which are the normal defaults,
  are also included for completeness.  (Tnx: Takeshi-Okazaki,
  MTSAT-1R Chief Aeronautical Mission Engineer)

* RS-232 output in PREDICT and Moontracker has been improved
  to prevent the transmission of extraneous bytes past the
  end-of-line characters.  (Tnx: SP7HAA)

* Since the bzero() function is being deprecated, all bzero()
  function calls were replaced by memset() in all affected code.

* The accuracy of PREDICT's real-time clock displays have been
  improved.

* A problem was corrected where the GET_SAT_POS socket command
  would return results as if the GET_SAT command were also
  sent to PREDICT.  (Tnx: David Hutchinson, G8SQH) 

* PREDICT now returns a response ("Huh?") to clients that send
  unrecognized commands via the UDP socket interface instead
  of sending nothing.

* The GET_DOPPLER socket command now returns Doppler shift
  information even if the satellite in question is below the
  local horizon.

* A keyboard interactive socket interface demonstration program
  was added ("demo-i" under clients/samples).

* "earthtrack" was updated to handle changes in xplanet version
  1.0 and later.

* A small bug in the orbital prediction [P] and visual calendar [V]
  modes of PREDICT was fixed.  Before the fix, the sunlight/visibility
  symbol (*, +) displayed in these modes would sometimes list the
  point of LOS as a '*' when it should have been a '+' at the
  conclusion of a visible pass.

* Array overflow conditions in predict.c's Print() and Daynum2String()
  functions were identified and fixed.

* Several very minor coding changes were made to allow warning-free
  compilation under GCC 4.x.x.

----------------------------------------------------------------------------

Release 2.2.2:
By John A. Magliacane <kd2bd@amsat.org> (14-Mar-2003):

* PREDICT's main menu was expanded to handle the additional program
  options.  The main menu also now displays the text "Server Mode"
  when the program is invoked in socket server mode.

* Orbital phase, eclipse depth, and squint angle data has been appended
  to the information returned to UDP client applications in response
  to the GET_SAT command.

* The Moon's declination, Greenwich hour angle, and right ascension
  have been appended to the information returned to UDP clients in
  response to the GET_MOON command.

* The Sun's declination, Greenwich hour angle, and right ascension
  have been appended to the information returned to UDP clients in
  response to the GET_SUN command.

* PREDICT now returns orbital data in NASA Two-Line format to UDP
  clients in response to the GET_TLE command.

* Added a new UDP socket command (GET_MODE) to return PREDICT's
  current tracking mode (single, multi, none) to networked clients.
  If PREDICT is operating in Single Satellite tracking mode, then the
  name of the satellite currently being tracked is returned by GET_MODE.

* Fixed a bug that caused GET_SAT requests to fail after GET_DOPPLER
  requests were issued. (Tnx: KF6RFX)

* Fixed a bug in "earthtrack" where latitude and longitude options sent
  to "xplanet" were reversed under certain circumstances.  (Tnx: ZL2BSJ)

* In Main Menu option [G] (Edit Ground Station Information), ground
  station latitude and longitude may now be entered in degree, minute,
  second (DMS) format in addition to decimal degrees.

* Moon tracking code replaced with a more precise algorithm based
  on Meeus' method.

* Added a lunar prediction mode that predicts passes of the Moon.

* Added a solar prediction mode that predicts passes of the Sun.

* Added "MoonTracker", a small background utility that directs
  an AZ/EL antenna array toward the position of the Moon.

* The "vocalizer" feature was greatly improved.  Sound samples are
  now read from ubiquitious .wav files, making it easier for users
  to create samples of their own, even in languages other than English.
  PREDICT now articulates numeric values in their entirety (ie:
  "three hundred forty nine") instead of articulating each digit
  individually ("three four nine") as was done previously.  Alarms
  and additional announcements were added to alert a visual observer
  when a spacecraft is visible, when it enters into sunlight, and
  when it enters into eclipse.

* Added "fodtrack", written by Luc Langermann, LX1GT.  This utility
  permits AZ/EL antenna control using the Fodtrack hardware interface
  via the parallel port.

* Several other coding changes were made.  Linux-specific code was
  removed in an effort to allow easier compilation under FreeBSD,
  MacOS X, and other Unixes in general.

* Program documentation was updated to reflect the changes and new
  features available in this version of PREDICT.

----------------------------------------------------------------------------

Release 2.2.1:
By John A. Magliacane <kd2bd@amsat.org> (14-Oct-2002):

* Fixed a problem that caused PREDICT to display the last satellite
  twice in listings when less than the maximum number of satellites
  existed in predict.tle.


installer.c
/*******************************************************/
/**       PREDICT installation program by KD2BD.      **/
/**   This program is compiled and executed through   **/
/**     the "configure" script in this directory.     **/
/**   It checks for the existence of a soundcard,     **/
/**  creates an appropriate "predict.h" file based    **/
/**  on the directory in which PREDICT was installed, **/
/**  compiles PREDICT and associated utilities, and   **/
/**    sets symbolic links between the executables    **/
/**     generated and the installation directory      **/
/**       specified (/usr/local/bin by default).      **/ 
/**                                                   **/
/**  Created: Oct 1999 -==- Last update: 14-May-2006  **/
/*******************************************************/

#include <curses.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <sys/soundcard.h>

char version[10];

void logo()
{
	attrset(COLOR_PAIR(6)|A_REVERSE|A_BOLD);
	mvprintw(3,18,"                                           ");
	mvprintw(4,18,"         --== PREDICT  v%s ==--         ",version);
	mvprintw(5,18,"   Released by John A. Magliacane, KD2BD   ");
	mvprintw(6,18,"                  May 2006                 ");
	mvprintw(7,18,"                                           ");
}

int main (argc,argv)
char argc, *argv[];
{
	int x, cc, dsp=-1;
	char pwd[80], src_path[255], dest_path[255], ans, destination[80];
	FILE *infile, *outfile;

	initscr();
	start_color();
	cbreak();
	scrollok(stdscr,TRUE);

	init_pair(1,COLOR_WHITE,COLOR_BLUE);
	init_pair(2,COLOR_RED,COLOR_BLUE);
	init_pair(3,COLOR_CYAN,COLOR_BLUE);
	init_pair(4,COLOR_GREEN,COLOR_BLUE);
	init_pair(5,COLOR_YELLOW,COLOR_BLUE);
	init_pair(6,COLOR_RED,COLOR_WHITE);

	cc=0;
	getcwd(pwd,79);
	dsp=open("/dev/dsp",O_WRONLY);

	if (dsp!=-1)
		close(dsp);

	infile=fopen(".version","r");
	fscanf(infile,"%s",version);
	fclose(infile);

	bkgdset(COLOR_PAIR(1)|A_BOLD);
	clear();
	refresh();
	logo();
	attrset(COLOR_PAIR(5)|A_BOLD);

	mvprintw(10,2,"PREDICT is a satellite tracking and orbital prediction program written for\n");
	printw("  Linux and similar operating systems by John A. Magliacane, KD2BD.\n");
	printw("  PREDICT is free software.  You can redistribute it and/or modify it under\n");
	printw("  the terms of the GNU General Public License as published by the Free\n");
	printw("  Software Foundation, either version 2 of the License or any later version.\n\n");
	printw("  PREDICT is distributed in the hope that it will useful, but WITHOUT ANY\n");
	printw("  WARRANTY, without even the implied warranty of MERCHANTABILITY or FITNESS\n");
	printw("  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more\n");
	printw("  details.\n\n");

	attrset(COLOR_PAIR(1)|A_BOLD);
	printw("  Do you accept these conditions and wish to install this software? [Y/N] ");
	refresh();

	do
	{
		ans=getch();

	} while (ans!='n' && ans!='N' && ans!='y' && ans!='Y' && ans!=27);


	if (ans=='y' || ans=='Y')
	{
		clear();
		curs_set(0);
		logo();
		attrset(COLOR_PAIR(4)|A_BOLD);
		printw("\n\n\n\n  PREDICT appears to be installed under %s/\n",pwd);

		if (dsp!=-1)
			printw("  An audio device was found at /dev/dsp");
		else
			printw("  No soundcard was found in your system... Bummer!");

		/* Write predict.h */

		outfile=fopen("predict.h","w");

		fprintf(outfile,"/* This file was generated by the installer program */\n\n");

		fprintf(outfile,"char *predictpath={\"%s/\"}, ",pwd);

		if (dsp==-1)
			fprintf(outfile, "soundcard=0,");
		else
			fprintf(outfile, "soundcard=1,");

		fprintf(outfile, " *version={\"%s\"};\n",version);

		fclose(outfile);

		printw("\n  predict.h was successfully created!\n");
		attrset(COLOR_PAIR(3)|A_BOLD);
		printw("\n  Now compiling PREDICT.  (This may take a while...)\n");
		refresh();

		/* Compile PREDICT... */

		cc=system("cc -Wall -O3 -s -fomit-frame-pointer -lm -lncurses -pthread predict.c -o predict");

		/* Create vocalizer.h */

		outfile=fopen("vocalizer/vocalizer.h","w");

		fprintf(outfile,"/* This file was generated by the installer program */\n\n");
		fprintf(outfile,"char *path={\"%s/vocalizer/\"};\n",pwd);

		fclose(outfile);

		if (dsp!=-1 && cc==0)
		{
			attrset(COLOR_PAIR(3)|A_BOLD);
			printw("  Compiling vocalizer...\n");
			refresh();
			system("cc -Wall -O3 -s -fomit-frame-pointer vocalizer/vocalizer.c -o vocalizer/vocalizer");
		}

		/* Now install the programs and man page by creating symlinks
		   between the newly created executables and the destination
		   directory.  The default destination directory for the
	   	   executables is /usr/local/bin.  This default may be
		   overridden by specifying a different path as an argument
		   to this program (ie: installer /usr/bin).  Normally this
		   is passed along from the "configure" script. */
	
		if (argc==2)
			strncpy(destination,argv[1],78);
		else	
			strncpy(destination,"/usr/local/bin/",15);

		/* Ensure a trailing '/' is
		   present in "destination". */

		x=strlen(destination);

		if (destination[x-1]!='/' && x!=0)
		{
			destination[x]='/';
			destination[x+1]=0;
		}

		if (cc==0)
		{
			attrset(COLOR_PAIR(3)|A_BOLD);
			printw("  Linking PREDICT binaries to %s\n\n",destination);
			sprintf(dest_path,"%spredict",destination);
			unlink(dest_path);
			sprintf(src_path,"%s/predict",pwd);
			symlink(src_path,dest_path);
			sprintf(dest_path,"%sxpredict",destination);
			unlink(dest_path);
			sprintf(src_path,"%s/xpredict",pwd);
			symlink(src_path,dest_path);
			unlink("/usr/local/man/man1/predict.1");
			sprintf(dest_path,"%s/docs/man/predict.1",pwd);
			symlink(dest_path,"/usr/local/man/man1/predict.1");

			attrset(COLOR_PAIR(5)|A_BOLD);
			printw("  Don't forget to check out the new graphical\n");
			printw("  client applications under the 'clients' directory!");
			attrset(COLOR_PAIR(1)|A_BOLD);
			printw("\n\n  Done!  Visit http://www.qsl.net/kd2bd/predict.html for the latest news!");
		}

		else
		{
			attrset(COLOR_PAIR(2)|A_BOLD);
			printw("  *** Compilation failed.  Program not installed.  :-(");
			beep();
		}
	}

	refresh();
	unlink("installer");
	curs_set(1);	
	refresh();
	endwin();
	exit(0);
}


vocalizer/vocalizer.c
/****************************************************************************
*                  VOCALIZER: A speech articulation program                 *
*		Copyright John A. Magliacane, KD2BD 1999-2003               *
*	   		  Last update: 02-Mar-2003                          *
*****************************************************************************
* This program is used by PREDICT to annunciate satellite coordinates       *
* through the system's soundcard (/dev/dsp) using .wav files containing     *
* uncompressed "raw" PCM data.  It is called as a background process.       *
*****************************************************************************
*                                                                           *
* This program is free software; you can redistribute it and/or modify it   *
* under the terms of the GNU General Public License as published by the     *
* Free Software Foundation; either version 2 of the License or any later    *
* version.                                                                  *
*                                                                           *
* This program is distributed in the hope that it will useful, but WITHOUT  *
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or     *
* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License     *
* for more details.                                                         *
*                                                                           *
*****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/soundcard.h>
#include "vocalizer.h"

int dsp;
char numstr[20][10];
unsigned char buffer[65536];

unsigned long buffer2long(indx)
int indx;
{
	unsigned long byte0, byte1, byte2, byte3;

	if ((indx+3)<65536)
	{
		byte0=(unsigned long)buffer[indx];
		byte1=(unsigned long)buffer[indx+1];
		byte2=(unsigned long)buffer[indx+2];
		byte3=(unsigned long)buffer[indx+3];

		return (byte0|(byte1<<8)|(byte2<<16)|(byte3<<24));
	}
	else
		return 0L;
}

unsigned int buffer2int(indx)
int indx;
{
	unsigned int byte0, byte1;

	if ((indx+1)<65536)
	{
		byte0=(unsigned int)buffer[indx];
		byte1=(unsigned int)buffer[indx+1];

		return (byte0|(byte1<<8));
	}
	else
		return 0;
}

int wavplay(filename)
char *filename;
{
	int x, y, format, bits, bytes, channels, block_size, dsp, fd;
	unsigned long total_samples=0L, rate=0L, running_total=0L;
	char filenpath[80];

	strncpy(filenpath,path,79);
	strncat(filenpath,filename,79);
	strncat(filenpath,".wav",79);

	fd=open(filenpath,O_RDONLY);

	if (fd==-1)
	{
		fprintf(stderr,"*** vocalizer: Cannot open \"%s\"\n",filenpath);
		return -1;
	}

	bytes=read(fd,&buffer,65536);

	y=((buffer[0]^'R')  | (buffer[1]^'I') |
  	   (buffer[2]^'F')  | (buffer[3]^'F') |
	   (buffer[8]^'W')  | (buffer[9]^'A') |
	   (buffer[10]^'V') | (buffer[11]^'E'));

	if (y)
	{
		fprintf(stderr,"*** vocalizer: No RIFF header found!\n");
		return -1;
	}

	for (x=12, y=1; x<65535 && y; x++)
		y=((buffer[x]^'f')  | (buffer[x+1]^'m') |
		  (buffer[x+2]^'t') | (buffer[x+3]^' '));

	if (y==0)
	{
		x+=3;
		x+=4;
		format=buffer2int(x);
		x+=2;
		channels=buffer2int(x);
		x+=2;
		rate=buffer2long(x);
		x+=4;
		x+=4;
		block_size=buffer2int(x);
		x+=2;
		bits=buffer2int(x);
		x+=2;

		if (format!=1)
		{
			fprintf(stderr,"*** vocalizer: Not PCM Wave Format.\n");
			return -1;
		}

		for (x=0, y=1; x<65535 && y; x++)
			y=((buffer[x]^'d') | (buffer[x+1]^'a') |
			 (buffer[x+2]^'t') | (buffer[x+3]^'a'));

		if (y==0)
		{
			x+=3;
			total_samples=buffer2long(x);

			if (total_samples!=0L)
			{
				dsp=open("/dev/dsp",O_RDWR);

				if (dsp==-1)
				{
					fprintf(stderr,"*** vocalizer: Cannot open /dev/dsp!\n");
					return -1;
				}

				ioctl(dsp,SOUND_PCM_WRITE_CHANNELS,&channels);
				ioctl(dsp,SOUND_PCM_WRITE_RATE,&rate);
				ioctl(dsp,SOUND_PCM_WRITE_BITS,&bits);

				lseek(fd,(long)(x+4),SEEK_SET);
				bytes=read(fd,&buffer,65536);
				running_total+=bytes;

				if (running_total>total_samples)
					bytes-=(running_total-total_samples);

				while (bytes>0)
				{
					write(dsp,&buffer,bytes);

					bytes=read(fd,&buffer,65536);
					running_total+=bytes;

					if (running_total>total_samples)
						bytes-=(running_total-total_samples);
				}

				close(dsp);
				close(fd);
			}
		}
		else
			fprintf(stderr,"*** vocalizer: Could not find 'data' chunk.\n");
	}
	else
		fprintf(stderr,"*** vocalizer: Could not extract format from header.\n");

	return 0;
}

void saynumber(num)
int num;
{
	char string[10];

	sprintf(string,"%4d",num);

	if (string[1]!=32)
	{
		wavplay(numstr[string[1]-48]);
		wavplay("hundred");
	}

	switch (string[2])
	{
		case '1':
			wavplay(numstr[(10+(string[3]-48))]);
			break;

		case '2':	
			wavplay("twenty");
			break;

		case '3':
			wavplay("thirty");
			break;

		case '4':
			wavplay("forty");
			break;

		case '5':
			wavplay("fifty");
			break;

		case '6':
			wavplay("sixty");
			break;

		case '7':
			wavplay("seventy");
			break;

		case '8':
			wavplay("eighty");
			break;

		case '9':
			wavplay("ninety");
			break;
	}

	if (string[3]!=32 && string[3]!=0 && string[2]!='1' && string[3]!='0')
		wavplay(numstr[string[3]-48]);

	if (string[3]=='0' && string[2]==32 && string[1]==32)
		wavplay("zero"); 
}

int main(argc,argv)
int argc;
char *argv[];
{
	int number;

	strcpy(numstr[1], "one\0");
	strcpy(numstr[2], "two\0");
	strcpy(numstr[3], "three\0");
     	strcpy(numstr[4], "four\0");
	strcpy(numstr[5], "five\0");
	strcpy(numstr[6], "six\0");
     	strcpy(numstr[7], "seven\0");
	strcpy(numstr[8], "eight\0");
	strcpy(numstr[9], "nine\0");
     	strcpy(numstr[0], "zero\0");
	strcpy(numstr[10],"ten\0");
	strcpy(numstr[11],"eleven\0");
	strcpy(numstr[12],"twelve\0");
	strcpy(numstr[13],"thirteen\0");
	strcpy(numstr[14],"fourteen\0");
	strcpy(numstr[15],"fifteen\0");
	strcpy(numstr[16],"sixteen\0");
	strcpy(numstr[17],"seventeen\0");
	strcpy(numstr[18],"eighteen\0");
	strcpy(numstr[19],"nineteen\0");

	/* Start talking! */

	if (argc==2)
	{
		if (strncmp(argv[1],"eclipse",7)==0)
		{
			wavplay("alarm");
			wavplay("eclipse");
		}

		if (strncmp(argv[1],"sunlight",8)==0)
		{
			wavplay("alarm");
			wavplay("sunlight");
		}

		if (strncmp(argv[1],"los",3)==0)
			wavplay("los");
	}

	if (argc>=3)
	{
		wavplay("intro");

		number=atoi(argv[1]);
		saynumber(number);
		wavplay("azimuth");

		number=atoi(argv[2]);
		saynumber(number);
		wavplay("elevation");

		if (argc>=4)
		{
			if (argv[3][0]=='+')
				wavplay("approaching");

			if (argv[3][0]=='-')
				wavplay("receding");
		}

		if (argc==5)
		{
			if (argv[4][0]=='V')
				wavplay("visible");
		}
	}

	return 0;
}
